### PR TITLE
[POC] better & easy multi themes

### PR DESCRIPTION
**Pull request by Its-Just-Nans**
# [POC] easy multi themes

## How to use

With this modification you can now you other languages to create your own theme !

You simply need to add the file `themeConfig.json` at the root of your theme
```json
{
    "type": "static",
    "expose": "/dist"
}
```
> Legend : 
>
> - put this file in `/themes/THEME_NAME/themeConfig.json`
> - the `type` is used by Aquila to know that the theme is not a NextJs theme
> - the `expose` attribute is used to expose the correct path in your theme folder

This configuration file will be saved (as usual) in the collection `themeConfig`

Note that after the installation, the theme is builded using `yarn build` or `yarn run build`, so if you need to create a build command, you can do it in `scripts` part of your `package.json`

## Tested with :

> ⚠️ Remember to add `themeConfig.json` with `type` different from `"next"`

### Vue app

Using
```
vue create hello-world
```
> From the example in https://cli.vuejs.org/guide/creating-a-project.html

After I installed the Vue app, I added the `themeConfig.json` with `"expose" : "/dist"`
With this test, I discover the issue in `know issue`

### Svelte app

Using 
```
npx degit sveltejs/template my-svelte-project
```
> From the example in https://svelte.dev/

After I installed the Svelte app, I added the `themeConfig.json` with `"expose" : "/public"`

## Know issues :
If you add your theme in production mode, the command `yarn install` will have only install dependencies and not devDependencies, so if your theme needs devDependencies to be builded, you need to be in developpement mode.

Because in developpment mode, the the command `yarn install` will install everything !

## Improvements

⚠️ ~~This improvement is not in the POC~~

Add a new property in `themeConfig.json` named `buildAtStart` like this :
```json
{
    "buildAtStart": true
}
```
> Legend : 
>
> - Like this, this options allow to build at start even in production mode !

### CODE REVIEW

**Comment by Its-Just-Nans** [comment]
- Support `buildAtStart` options
- Add new button in backoffice to build the theme

**Comment by Its-Just-Nans** [comment]
Futur possible improvements :
- remove all thing that are linked with the `default_theme` and put these things in files IN the theme folder

**Comment by Its-Just-Nans** [comment]
Now using `utils/themes.js` to use `yarn` commands

**Comment by Its-Just-Nans** [comment]
## Tested with Angular

Using
```sh
npm install -g @angular/cli
ng new my-app
```

Then I added the file `themeConfig.json`
```json

```
The `npm run build` produce `dist/my-app` so you have two options to expose the website correctly :
- change the angular configuration from `"outputPath": "dist/my-app",` to  `"outputPath": "dist/"`
- change the `expose` attribute of the `themeConfig.json` file to "dist/my-app"

Have fun ! ;)

**Comment by Its-Just-Nans** [comment]
# POC - V2

Now themes are separate from AquilaCMS.

Introducing two new features :
- custom build for theme
- global hook to implements at start of Aquila (for example, to use server side rendering)

Introducing new file :
- `themeInit.js`

## Custom build for theme

To build your theme you have now two options :
- custom building with a function
- custom building with a command

### Custom building with a function

You just need to export a function named `build` in your `themeInit.js`

Example of `themeInit.js`
```js
const build = async () => {
  // here I can do build
};

module.exports = {
    start,
    build
};
```

### Custom building with a command

If the function `build` is not defined, then AquilaCMS will simply run the `yarn run build` command of the `package.json` of your theme

In that case, you can create your command in your `package.json`

## global hook to implements at start of AquilaCMS

If your theme don't have any `type`, AquilaCMS will chose the default `type` custom.

This type allow you to add a hook in the `server.js`, with the function `start` of your `themeInit.js`

Example for start function
```js
const start = async (server) => {
  // do thing, add handler, do what you want
    return handler; // optionnal
}
module.exports = {
    start
};
```
> Legend :
>
> - As you can see, the function start take the express server in parameters, like this, you can add handler to the server
> - As you can see, the function start can return a last handler, to be used in express

## Introducing new theme building:

The AquilaCMS command `yarn run build` as changed to a script.
This script compile the theme (with the name given as paramter) and if there is a custom build, it use the custom build

**Comment by Its-Just-Nans** [comment]
Tested and working with NuxtJS

SSR && static

File: themes/README.md [hunk 1]
```diff
@@ -1,3 +1,23 @@
...
+
+Theme can use options in `themeConfig.json`
+
+Options are listed here:
+
+```json
+{
+    "type": "static",
+    "expose": "/dist",
+    "buildAtStart": true
```
Comment (devaquila) [inline themes/README.md:15]: It's not the theme (themeConfig.json) that decides this, but Aquila

File: themes/README.md [hunk 2]
```diff
@@ -1,3 +1,23 @@
-Les fichiers des thèmes doivent être placés dans ce répertoire
+# Themes
 
-The theme files must be placed in this directory
+Theme folders must be placed in this directory
+
+## Options
+
+Theme can use options in `themeConfig.json`
```
Comment (devaquila) [inline themes/README.md:7]: Even VueJS's themes ?

File: themes/default_theme/themeInit.js [hunk 3]
```diff
@@ -0,0 +1,86 @@
+const fs = require('fs');
```
Comment (devaquila) [inline themes/default_theme/themeInit.js:1]: Every themes need this file ?
Maybe explains this in the readme (in /theme)
Any sample for VueJS or Angular ?

File: utils/build.js [hunk 4]
```diff
@@ -0,0 +1,31 @@
+const path        = require('path');
```
Comment (devaquila) [inline utils/build.js:1]: Move this in /utils/themes.js ?

File: utils/build.js [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+const fs          = require('fs');
+const utilsThemes = require('./themes');
+
+/**
+ * ONLY npm run build "theme"
+ */
+/* eslint-disable */
+const buildOneTheme = async () => {
```
Comment (devaquila) [inline utils/build.js:9]: Is buildOneTheme works with vuejs?

File: utils/themes.js [hunk 6]
```diff
@@ -78,9 +75,43 @@ const yarnBuild = async (themeName = '') => {
     return returnValues;
 };
 
+/**
+ * @description setConfigTheme
+ * @param theme : String Theme selectionné
+ */
+const setConfigTheme = async (theme) => {
```
Comment (devaquila) [inline utils/themes.js:82]: Is there really usefull ? (themeConfig.json)

File: services/themes.js [hunk 7]
```diff
@@ -105,29 +111,6 @@ const uploadTheme = async (originalname, filepath) => {
...
-        if (oldConfig) {
-            const mergedConfig = {...config, ...oldConfig.config}; // We merge the old and the new configuration to not lose the data
-            await ThemeConfig.updateOne({name: theme}, {$set: {name: theme, config: mergedConfig}});
-        } else {
-            await ThemeConfig.create({name: theme, config});
-        }
-    } catch (err) {
-        // nothing
-    }
-};
```
Comment (Its-Just-Nans) [inline services/themes.js:129]: deleted (deprecated)

File: backoffice/app/themes/themes.controllers.js [hunk 8]
```diff
@@ -53,6 +47,18 @@ ThemesController.controller("ThemesCtrl", [
...
+            if (!appUrl) {
+                correctAppUrl = "/";
+            } else if (!appUrl.endsWith("/")) {
+                correctAppUrl = `${appUrl}/`;
+            } else {
+                correctAppUrl = appUrl;
+            }
+            return correctAppUrl + adminPrefix;
+        }
+
```
Comment (Its-Just-Nans) [inline backoffice/app/themes/themes.controllers.js:61]: moved function to hava "global" access

File: backoffice/app/themes/themes.controllers.js [hunk 9]
```diff
@@ -108,13 +114,31 @@ ThemesController.controller("ThemesCtrl", [
...
+                        if ($scope.progressValue == 100) {
+                            setTimeout(function () {
+                                location.href = window.location = buildAdminUrl($scope.config.environment.appUrl, $scope.config.environment.adminPrefix);
+                            }, 7000);
+                        }
+
+                        if ($scope.progressValue >= 110) {
+                            $interval.cancel(timerRestart);
+                        }
+                    }, 250);
```
Comment (Its-Just-Nans) [inline backoffice/app/themes/themes.controllers.js:141]: Restart after build
(there was a restart but no informations on the front :( )

**Pull request merged**